cunumeric/deferred.py

```diff
@@ -305,37 +305,278 @@ def get_scalar_array(self):
+        # NumPy array.
+        N = self.ndim
+        pointN_dtype = self.runtime.get_point_type(N)
+        store = self.context.create_store(
+            pointN_dtype, shape=out_shape, optimize_scalar=True
+        )
+        output_arr = DeferredArray(
+            self.runtime, base=store, dtype=pointN_dtype
+        )
+
+        # call ZIP function to combine index arrays into a singe array
+        task = self.context.create_task(CuNumericOpCode.ZIP)
+        task.add_output(output_arr.base)
+        if len(arrays) < self.ndim:
+            task.add_scalar_arg(self.ndim, ty.int64)  # N of points in Point<N>
+            task.add_scalar_arg(key_dim, ty.int64)  # key_dim
+            task.add_scalar_arg(start_index, ty.int64)  # start_index
+            for a in arrays:
+                task.add_input(a)
+                task.add_alignment(output_arr.base, a)
+                task.add_broadcast(a, axes=tuple(range(1, len(out_shape))))
+                task.add_broadcast(
+                    output_arr.base, axes=tuple(range(1, len(out_shape)))
+                )
+        else:
+            task.add_scalar_arg(self.ndim, ty.int64)
+            task.add_scalar_arg(self.ndim, ty.int64)
+            task.add_scalar_arg(start_index, ty.int64)
+            for index_arr in arrays:
+                task.add_input(index_arr)
+                task.add_alignment(output_arr.base, index_arr)
+        task.execute()
+
+        return output_arr
+
+    def _copy_store(self, store):
+        store_to_copy = DeferredArray(
+            self.runtime,
+            base=store,
+            dtype=self.dtype,
+        )
+        store_copy = self.runtime.create_empty_thunk(
+            store_to_copy.shape,
+            self.dtype,
+            inputs=[store_to_copy],
+        )
+        store_copy.copy(store_to_copy, deep=True)
+        return store_copy, store_copy.base
+
+    def _create_indexing_array(self, key, is_set=False):
+        store = self.base
+        rhs = self
+        # the index where the first index_array is passed to the [] operator
+        start_index = -1
         if isinstance(key, tuple):
+            key = self._unpack_ellipsis(key, self.ndim)
+            shift = 0
+            last_index = self.ndim
+            # in case when index arrays are passed in the scaterred way,
+            # we need to transpose original array so all index arrays
+            # are close to each other
+            transpose_needed = False
+            transpose_indices = tuple()
+            # since we can't call Copy operation on transformed Store, after
+            # the transformation, we need to return a copy
             tuple_of_arrays = ()
-            for k in key:
-                if not isinstance(k, NumPyThunk):
-                    raise NotImplementedError(
-                        "need support for mixed advanced indexing"
+            index_map = []
+
+            # First, we need to check if transpose is needed
```

Comment:
This doesn't work correctly with boolean arrays. For example, in this case:

```
x = cn.ones((2,3,4,5,3,4))
ind1 = cn.full((3,4), True)
ind2 = cn.full((3,4), True)
print(x[:,ind1,:,ind2].shape)
```

The value of `transpose_indices` is `(1, 3, 0, 2, 4, 5)`, but it should be `(1, 2, 4, 5, 0, 3)` (i.e. the dimensionality of the boolean arrays should be considered).